FAB_LED.h

```diff
@@ -885,100 +885,129 @@ avrBitbangLedStrip<FAB_TVAR>::twoPortSoftwareSendBytes(const uint16_t count, con
-				FAB_PORT(dataPortId, 0xFF);
-				DELAY_CYCLES(high0 - sbiCycles);
-
-				FAB_PORT(dataPortId, bitmask);
-				DELAY_CYCLES(high1 - sbiCycles - high0);
-
-				FAB_PORT(dataPortId, 0x00);
-				// Let's assume we'll spend enough time doing math to not need to wait
-				//DELAY_CYCLES(low0 - cbiCycles);
-			}
+	const uint16_t blockSize = count / 8;
+	uint8_t r0 __asm__("r2") = 0;
+	uint8_t r1 __asm__("r3") = 0;
+	uint8_t r2 __asm__("r4") = 0;
+	uint8_t r3 __asm__("r5") = 0;
+	uint8_t r4 __asm__("r6") = 0;
+	uint8_t r5 __asm__("r7") = 0;
+	uint8_t r6 __asm__("r8") = 0;
+	uint8_t r7 __asm__("r9") = 0;
+
+	for(register uint16_t c = 0; c < blockSize; c++) {
+/*
+		r0 = array[c];
+		r1 = array[c + 1 * blockSize];
+		r2 = array[c + 2 * blockSize];
+		r3 = array[c + 3 * blockSize];
+*/
+		r4 = array[c + 4 * blockSize];
+		r5 = array[c + 5 * blockSize];
+		r6 = array[c + 6 * blockSize];
+		r7 = array[c + 7 * blockSize];
+		for(register int8_t b = 7; b >= 0; b--) {
+			uint8_t bitmask __asm__("r10") = 0;
+
+			bitmask |=  r0 & 1      ; r0 >>= 1;
+			bitmask |= (r1 & 1) << 1; r1 >>= 1;
+			bitmask |= (r2 & 1) << 2; r2 >>= 1;
+			bitmask |= (r3 & 1) << 3; r3 >>= 1;
+			bitmask |= (r4 & 1) << 4; r4 >>= 1;
+			bitmask |= (r5 & 1) << 5; r5 >>= 1;
+			bitmask |= (r6 & 1) << 6; r6 >>= 1;
+			bitmask |= (r7 & 1) << 7; r7 >>= 1;
+
+			// Set all HIGH, set LOW all zeros, set LOW zeros and ones.
+			FAB_PORT(dataPortId, 0xFF);
+			DELAY_CYCLES(2); //high0 - sbiCycles);
+
+//			FAB_PORT(dataPortId, bitmask);
+			AVR_PORT(dataPortId) = bitmask;
+			DELAY_CYCLES(6); //high1 - sbiCycles - high0);
+
+			FAB_PORT(dataPortId, 0x00);
+			// Let's assume we'll spend enough time doing math to not need to wait
+			DELAY_CYCLES(10); //low0 - cbiCycles);
 		}
 	}
 }
+#endif
 
-#if 0
 template<FAB_TDEF>
 inline void
 avrBitbangLedStrip<FAB_TVAR>::eightPortSoftwareSendBytes(const uint16_t count, const uint8_t * array)
 {
-	const uint16_t blockSize = count /2;
+	const uint16_t blockSize = count / 8;
+	uint8_t r0 __asm__("r2") = 0;
+	uint8_t r1 __asm__("r3") = 0;
+	uint8_t r2 __asm__("r4") = 0;
+	uint8_t r3 __asm__("r5") = 0;
+	uint8_t r4 __asm__("r6") = 0;
+	uint8_t r5 __asm__("r7") = 0;
+	uint8_t r6 __asm__("r8") = 0;
+	uint8_t r7 __asm__("r9") = 0;
+
+	for(uint16_t c __asm__("r11") = 0; c < blockSize; c++) {
+		for(int8_t b __asm__("r12") = 7; b >= 0; b--) {
+			uint8_t bitmask __asm__("r10") = 0;
+
+			bitmask |=  r0 & 1      ; r0 >>= 1;
+			bitmask |= (r1 & 1) << 1; r1 >>= 1;
+			bitmask |= (r2 & 1) << 2; r2 >>= 1;
+			bitmask |= (r3 & 1) << 3; r3 >>= 1;
+			bitmask |= (r4 & 1) << 4; r4 >>= 1;
+			bitmask |= (r5 & 1) << 5; r5 >>= 1;
+			bitmask |= (r6 & 1) << 6; r6 >>= 1;
+			bitmask |= (r7 & 1) << 7; r7 >>= 1;
+
+			// Load ONE byte per iteration
+			// Skip end condition check to reduce CPU cycles.
+			// It is expected that the LED strip won't have extra pixels.
+			if ((b >= dataPortPin) && (b <= clockPortPin)) switch (b) {
+				case 0:
+					r0 = array[c];
```

Comment:
You are correct. And that is the intent, the trick is when the byte is overwritten:

My goal is to pipeline using registers so that I read only ONE memory location per iteration. So every time I toggle the port, it resets my timeout, and I spend that time doing some register arithmetic and one memory access.

Now, the trick is: The different ports don't care when they get their bits pushed, so this loop delays each port by 1 cycle. Port 7 receives its first bit 8 cycles after port 0 got its first bit, effectively I push at time t=7 D7.bit0, D6.bit1... D0.bit7. Makes sense?

Now the kicker is: The way I wrote the code it doesn't seem like the code is compiled to do that but in fact behaves pretty much as if I was reading all 8 memory locations at the same cycle. And it seems I can't read much more than 4 of them at 16MHz.

So my pipeline code is not working as expected.